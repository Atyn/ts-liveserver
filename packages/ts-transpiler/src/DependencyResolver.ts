import Path from 'path'
import Resolve from 'enhanced-resolve'
import IEsmDependencyResolver from './types/IEsmDependencyResolver'

type ReturnType = {
	path: string
	ignore?: boolean
}

const RESOLVE_EXTENSIONS = ['.js', '.ts', '.tsx', '.jsx', '.json', '.mjs']
const ALIAS_FIELDS = ['browser', 'module']
export default class DependencyResolver implements IEsmDependencyResolver {
	private resolver = Resolve.create.sync({
		mainFields: [...ALIAS_FIELDS, 'main'],
		extensions: RESOLVE_EXTENSIONS,
		aliasFields: ALIAS_FIELDS,
	})
	private alias?: Record<string, string>
	constructor(alias: Record<string, string> = {}) {
		this.alias = alias
	}
	// Return e.g. ./hello/module.js
	public resolveDependencyName(
		parentFilePath: string,
		dependencyName: string,
	): ReturnType {
		if (
			dependencyName.startsWith('/') ||
			dependencyName.startsWith('https://') ||
			dependencyName.startsWith('http://')
		) {
			return {
				path: dependencyName,
			}
		}
		const resolveResult = this.resolveDependencyPath(
			parentFilePath,
			dependencyName,
		)
		if (resolveResult.ignore) {
			return resolveResult
		}
		const absolutePath = resolveResult.path
		const pathObj = Path.parse(absolutePath)
		const relativeDir =
			Path.relative(Path.dirname(parentFilePath), pathObj.dir) || '.'
		const result = relativeDir + '/' + pathObj.name + '.js'
		const posixResult = Path.posix.normalize(result).replace(/\\/g, '/')
		return {
			path: posixResult.startsWith('.') ? posixResult : './' + posixResult,
		}
	}
	// Return an absolute path e.g. /tmp/a-apath/node_modules/hello/module.js
	public resolveDependencyPath(
		parentFilePath: string,
		dependencyName: string,
	): ReturnType {
		if (this.alias && dependencyName in this.alias) {
			return this.resolveDependencyPath(
				parentFilePath,
				this.alias[dependencyName],
			)
		}
		const result = this.resolver(
			{},
			Path.dirname(parentFilePath),
			dependencyName,
		)
		if (result === false) {
			return {
				ignore: true,
				path: dependencyName,
			}
		}
		return {
			path: result,
		}
	}
}
