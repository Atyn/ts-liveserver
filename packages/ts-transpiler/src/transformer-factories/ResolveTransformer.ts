import TypeScript from 'typescript'
import Path from 'path'

export default class ResolveTransformer implements TypeScript.CustomTransformer {
  private context: TypeScript.TransformationContext
  constructor(context: TypeScript.TransformationContext) {
  	this.context = context
  }
  private resolvePath(a: string, b: string) {
  	const filePath = a
  	const dependencyFilePath = require.resolve(b, {
  		paths: [Path.dirname(filePath)],
  	})
  	const relativeDir =
      Path.relative(
      	Path.dirname(filePath),
      	Path.dirname(dependencyFilePath),
      ) || '.'
  	const relativeDependencyFilePath =
      relativeDir + '/' + Path.basename(dependencyFilePath)
  	return relativeDir + '/' + Path.basename(relativeDependencyFilePath)
  }
  private visit(node: TypeScript.Node) {
  	if (
  		TypeScript.isStringLiteral(node) &&
      TypeScript.isImportDeclaration(node.parent)
  	) {
  		console.log(node.getSourceFile().fileName, node.text);
  		return TypeScript.factory.createStringLiteral(
  			this.resolvePath(node.getSourceFile().fileName, node.text)
  		)
  	}
  	return TypeScript.visitEachChild(node, this.visit.bind(this), this.context)
  }
  transformSourceFile(node: TypeScript.SourceFile): TypeScript.SourceFile {    
  /*
    console.log(
  		node.resolvedModules
    )
    */
  	return TypeScript.visitNode(node, this.visit.bind(this))
  }
  transformBundle(node: TypeScript.Bundle): TypeScript.Bundle {
  	return node
  }
}
