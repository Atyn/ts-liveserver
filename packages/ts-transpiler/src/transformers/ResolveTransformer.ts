import TypeScript from 'typescript'
import Path from 'path'
import FS from 'fs'
/*
class ModuleResolutionHost implements TypeScript.ModuleResolutionHost {
	fileExists(fileName: string): boolean
	readFile(fileName: string): string | undefined
	trace?(s: string): void
	directoryExists?(directoryName: string): boolean
	realpath?(path: string): string
	getCurrentDirectory?(): string
	getDirectories?(path: string): string[]
}
*/

/* eslint-disable no-console*/

export default class ResolveTransformer
	implements TypeScript.CustomTransformer {
	private context: TypeScript.TransformationContext
	private moduleResolutionHost: TypeScript.ModuleResolutionHost
	constructor(context: TypeScript.TransformationContext) {
		this.context = context
		this.moduleResolutionHost = TypeScript.createCompilerHost(
			this.context.getCompilerOptions(),
		)
	}
	// Return a relative path to the module
	private resolveDependencyPath(
		parentPath: string,
		dependencyName: string,
	): string {
        const parentDirectory = Path.dirname(parentPath)

		if(dependencyName.startsWith('.')) {
            if(dependencyName.endsWith('.js') || dependencyName.endsWith('.json')) {
                return dependencyName;
            }
            return dependencyName + '.js';
        }

		return Path.relative(parentDirectory, require.resolve(dependencyName, {
            paths: this.getResolvePaths(parentDirectory),
        }))
	}
    // Returns paths used for resolving with require.resolve() (same method that Node.js is using for modules)
    private getResolvePaths(directory: string): string[] {
        const paths = [];
        const directories = directory.split(Path.sep);
        while(directories.length) {
            paths.push(directories.join(Path.sep) + Path.sep + 'node_modules');
            directories.pop();
        }
        return paths;
    }
	private visit(node: TypeScript.Node) {
		if (
			node.parent &&
			TypeScript.isStringLiteral(node) &&
			(TypeScript.isExportDeclaration(node.parent) ||
				TypeScript.isImportDeclaration(node.parent))
		) {
			return TypeScript.factory.createStringLiteral(
				this.resolveDependencyPath(node.getSourceFile().fileName, node.text),
			)
		}
		return TypeScript.visitEachChild(node, this.visit.bind(this), this.context)
	}
	transformSourceFile(node: TypeScript.SourceFile): TypeScript.SourceFile {
		return TypeScript.visitNode(node, this.visit.bind(this))
	}
	transformBundle(node: TypeScript.Bundle): TypeScript.Bundle {
		return node
	}
}
