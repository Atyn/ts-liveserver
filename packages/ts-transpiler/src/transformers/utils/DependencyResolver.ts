import TypeScript from 'typescript'
import Path from 'path'

export default class DependencyResolver {
	private context: TypeScript.TransformationContext
	private fileName: string
	constructor(fileName: string, context: TypeScript.TransformationContext) {
		this.fileName = fileName
		this.context = context
	}
	public resolveRelativeDependency(dependencyFileName: string): string {
		return this.resolveDependencyName(this.fileName, dependencyFileName)
	}

	// Return e.g. ./hello/module.js
	private resolveDependencyName(
		parentPath: string,
		dendencyName: string,
	): string {
		const absolutePath = this.resolveDependencyPath(parentPath, dendencyName)
		const pathObj = Path.parse(absolutePath)
		const relativeDir =
			Path.relative(Path.dirname(parentPath), pathObj.dir) || '.'
		const result = relativeDir + '/' + pathObj.name + pathObj.ext
		if (result.startsWith('.')) {
			return result.replace(/\.(ts|tsx|jsx)$/, '.js')
		}
		return './' + result.replace(/\.(ts|tsx|jsx)$/, '.js')
	}
	// Return an aboslute path e.g. /tmp/a-apath/node_modules/hello/module.js
	private resolveDependencyPath(
		parentPath: string,
		dendencyName: string,
	): string {
		const resolveResults = TypeScript.resolveModuleName(
			dendencyName,
			parentPath,
			this.context.getCompilerOptions(),
			TypeScript.createCompilerHost(this.context.getCompilerOptions()),
		)
		if (resolveResults?.resolvedModule?.isExternalLibraryImport) {
			const nodeResolve = require.resolve(dendencyName, {
				paths: [Path.dirname(parentPath)],
			})
			if (nodeResolve) {
				return nodeResolve
			}
		}
		const resolvedFileName = resolveResults?.resolvedModule?.resolvedFileName
		if (!resolvedFileName) {
			throw new Error(
				'Could not resolve' + dendencyName + 'from module' + parentPath,
			)
		}
		return resolvedFileName
	}
}
