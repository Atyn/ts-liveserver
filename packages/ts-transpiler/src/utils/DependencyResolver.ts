import Path from 'path'
import Resolve from 'enhanced-resolve'

const RESOLVE_EXTENSIONS = ['.js', '.ts', '.tsx', '.jsx', '.json', '.mjs']

export default class DependencyResolver {
	private resolver = Resolve.create.sync({
		mainFields: ['browser', 'module', 'main'],
		extensions: RESOLVE_EXTENSIONS,
		aliasFields: ['browser'],
	})
	private alias?: Record<string, string>
	constructor(alias: Record<string, string> = {}) {
		this.alias = alias
	}
	// Return e.g. ./hello/module.js
	public resolveDependencyName(
		parentFilePath: string,
		dependencyName: string,
	): string {
		const absolutePath = this.resolveDependencyPath(
			parentFilePath,
			dependencyName,
		)
		const pathObj = Path.parse(absolutePath)
		const relativeDir =
			Path.relative(Path.dirname(parentFilePath), pathObj.dir) || '.'
		const result = relativeDir + '/' + pathObj.name + '.js'
		const posixResult = Path.posix.normalize(result).replace(/\\/g, '/')
		return posixResult.startsWith('.') ? posixResult : './' + posixResult
	}
	// Return an aboslute path e.g. /tmp/a-apath/node_modules/hello/module.js
	public resolveDependencyPath(
		parentFilePath: string,
		dependencyName: string,
	): string {
		if (this.alias && dependencyName in this.alias) {
			return this.resolveDependencyPath(
				parentFilePath,
				this.alias[dependencyName],
			)
		}
		const directory = Path.dirname(parentFilePath)
		const result = this.resolver({}, directory, dependencyName)
		if (result === false) {
			throw new Error(
				'Could not resolve ' + dependencyName + ' from ' + parentFilePath,
			)
		}
		return result
	}
}
