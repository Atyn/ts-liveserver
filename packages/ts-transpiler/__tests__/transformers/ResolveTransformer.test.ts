import ResolveTransformer from '../../src/transformers/ResolveTransformer'
import DependencyResolver from '../../src/utils/DependencyResolver'
import TypeScript from 'typescript'

describe('ResolveTransformer', () => {
	beforeEach(() => {
		jest
			.spyOn(DependencyResolver.prototype, 'resolveRelativeDependency')
			.mockImplementation(() => 'B')
	})
	describe('transformSourceFile', () => {
		it('Should convert static import', async () => {
			const input = 'import "./hello.ts";'
			const output = 'import "B";'
			expect(await transformWithPlugin(input)).toBe(
				await transformWithoutPlugin(output),
			)
		})
		it('Should convert async import', async () => {
			const input = '{ import("./hello.ts"); }'
			const output = '{ import("B"); }'
			expect(await transformWithPlugin(input)).toBe(
				await transformWithoutPlugin(output),
			)
		})
	})
})

const compilerOptions: TypeScript.CompilerOptions = {
	allowJs: true,
	jsx: TypeScript.JsxEmit.React,
	checkJs: false,
	noResolve: true,
	esModuleInterop: true,
	skipLibCheck: true,
	pretty: true,
	allowUnreachableCode: true,
	target: TypeScript.ScriptTarget.ES2020,
	declaration: false,
	module: TypeScript.ModuleKind.ES2020,
	moduleResolution: TypeScript.ModuleResolutionKind.NodeJs,
	sourceMap: false,
}

const transformers: TypeScript.CustomTransformers = {
	after: [(context) => new ResolveTransformer(context)],
}

async function transformWithPlugin(code: string): Promise<string> {
	const results = await TypeScript.transpileModule(code, {
		compilerOptions: compilerOptions,
		fileName: 'hello.ts',
		transformers: transformers,
	})
	return results.outputText.trim()
}

async function transformWithoutPlugin(code: string): Promise<string> {
	const results = await TypeScript.transpileModule(code, {
		compilerOptions: compilerOptions,
		fileName: 'hello.ts',
	})
	return results.outputText.trim()
}
