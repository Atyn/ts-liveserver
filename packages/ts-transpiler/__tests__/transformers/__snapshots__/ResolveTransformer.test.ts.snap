// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResolveTransformer transformSourceFile Should convert async import 1`] = `
"{
    import(\\"B\\");
}"
`;

exports[`ResolveTransformer transformSourceFile Should convert forward 1`] = `"export * from \\"B\\";"`;

exports[`ResolveTransformer transformSourceFile Should convert named imports 1`] = `
"import { hello } from \\"B\\";
console.log(hello);"
`;

exports[`ResolveTransformer transformSourceFile Should convert static import 1`] = `"import \\"B\\";"`;
